### PR TITLE
ISSUE-29690: Fixing issue when the images' sort was using ID instead of position field

### LINKED ISSUE #29690: Magento 2 product YouTube video is not playing for the simple products within a configurable product
This issue was reported here ( https://github.com/magento/magento2/issues/23107 ) but could not be reproduced. In my tests in clean installations of v2.3.5-p1 this problem consistently happened when a video is added to the simple products within a configurable product if those simple products already have a photo and the video is moved to the position before (to the left of) image.

### Preconditions (*)

1. Magento 2.4-develop

### Steps to reproduce (*)

1. Create a configurable product containing 3 simple products that vary in color (make sure to upload an image for each simple product and for the configurable product, then save the configurable product).
2. Now that the listing is published, add a YouTube video to each simple product and move it to the first position (to the left of the image). Then save the changes.

### Expected result (*)

1. The video of any options (i.e. simple products) within the configurable product would be displayed.

### Actual result (*)

1. If you check the frontend and switch between the options (i.e. simple products), you will see that the video is just a thumbnail and will not play.
Peek 2020-09-03 15-23

The problem seems to be that the order of adding videos and images matters. The videos should be uploaded first, then the images should be added. Otherwise the videos would not play.

Workaround:
1. Remove the images from the simple products, so that only the video remains, then save the simple products.
2. Now add the images back to the simple products.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by viniciusbordinhao-blueacorn**
### Description (*)

The issue was happening because the images were ordered by ID instead of position. Due to that some items that were supposed to be videos was interpreted like image.
### Fixed Issues (if relevant)
1. Fixes magento/magento2#29690 

### Manual testing scenarios (*)
1. See #29690 
2. `./vendor/bin/phpunit app/code/Magento/ConfigurableProduct/Test/Unit/Block/Plugin/Product/Media/GalleryTest.php`

### Questions or comments
Those changes just fix the issues when the video is assigned to the main variant product, in other words, the first product related to the size.  For instance, when the user clicks on a different colour, even if the product has the video, it is not showing. I think it is another issue related to the javascript, that I will fix soon and send a PR

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by viniciusbordinhao-blueacorn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @viniciusbordinhao-blueacorn,
Could you cover your change with MFTF test as unit test isn't actually checking that issue isn't reproducing anymore?

**Comment by viniciusbordinhao-blueacorn** [comment]
@ihor-sviziev 

Sure, I will assign the task to me again. The Mtft tests will take some time but I will try to finish it next week.

* EE tests and B2B tests are breaking, however, I do not even have those tests in my local environment. I am not sure how I could fix it.

**Comment by viniciusbordinhao-blueacorn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by viniciusbordinhao-blueacorn** [comment]
@ihor-sviziev 

The tests have been broken due to the Youtube API key missing. How could I fix it? I think the .credentials file should have a Youtube key, am I right?

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Comment by viniciusbordinhao-blueacorn** [comment]
@ihor-sviziev 

The test has been breaking for the same reason of last execution:

c507cf1d903d678a

**Comment by ihor-sviziev** [comment]
@engcom-Charlie @engcom-Hotel could you help with failing functional tests?

**Comment by engcom-Hotel** [comment]
I'll help.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE

**Comment by ihor-sviziev** [comment]
@engcom-Hotel can we actually add skip for MFTF  test?

**Comment by engcom-Hotel** [comment]
@ihor-sviziev as it was discussed, for some tests it is possible on a temporary basis.

File: app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php [hunk 1]
```diff
@@ -59,7 +59,7 @@ public function afterGetOptionsMediaGalleryDataJson(
     private function getProductGallery($product)
```
Comment (ihor-sviziev) [inline app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php]: Please define argument & return types, like this:
```suggestion
    private function getProductGallery(Product $product): array
```

File: app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php [hunk 2]
```diff
@@ -69,4 +69,18 @@ private function getProductGallery($product)
         }
         return $result;
     }
+
+    /**
+     * @param Product $product
+     * @return array
+     */
+    private function getImagesOrderedByPosition($product)
```
Comment (ihor-sviziev) [inline app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php]:
```suggestion
    private function getImagesOrderedByPosition(Product $product): array
```

File: app/code/Magento/ConfigurableProduct/Test/Unit/Block/Plugin/Product/Media/GalleryTest.php [hunk 3]
```diff
@@ -41,13 +42,20 @@ public function testAfterGetOptions()
             ['media_type' => 'type', 'video_url' => 'url', 'file' => 'image.jpg']
         );
 
+        $dataCollection = $this->getMockBuilder(DataCollection::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['getItems'])
+            ->getMock();
+
+
```
Comment (ihor-sviziev) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Block/Plugin/Product/Media/GalleryTest.php]: No need to have there two empty lines
```suggestion

```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php (hunk 1 above)
Comment (engcom-Hotel) [reply app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php]: Has been updated

File: app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php (hunk 2 above)
Comment (engcom-Hotel) [reply app/code/Magento/ConfigurableProduct/Block/Plugin/Product/Media/Gallery.php]: Has been updated

File: app/code/Magento/ConfigurableProduct/Test/Unit/Block/Plugin/Product/Media/GalleryTest.php (hunk 3 above)
Comment (engcom-Hotel) [reply app/code/Magento/ConfigurableProduct/Test/Unit/Block/Plugin/Product/Media/GalleryTest.php]: Has been removed

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8532](https://jira.corp.magento.com/browse/ENGCOM-8532) has been created to process this Pull Request

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
**Manual testing sceanrio:**

- Create a configurable product containing 3 simple products that vary in color: 
- - make sure to upload an image for each simple product and for the configurable product, then save the configurable product
image

- Now that the listing is published, add a YouTube video to each simple product and move it to the first position (to the left of the image). 
- Save product
- Go to storefront
- Open product detail page

**Before:**
:x: The video is just a thumbnail and cannot be played
Peek 2020-09-03 15-23

**After:**
:heavy_check_mark: The video of any options (i.e. simple products) within the configurable product is displayed.
Peek 2020-12-15 19-39

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Pull request merged**